File: problems/3606-coupon-code-validator/solution.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<String> validateCoupons(String[] code, String[] businessLine, boolean[] isActive) {
        List<String> validCoupons = new ArrayList<>();
        Map<String, Integer> businessLineOrder = new HashMap<>();
        businessLineOrder.put("electronics", 1);
        businessLineOrder.put("grocery", 2);
        businessLineOrder.put("pharmacy", 3);
        businessLineOrder.put("restaurant", 4);

        List<Coupon> coupons = new ArrayList<>();

        for (int i = 0; i < code.length; i++) {
            if (isValidCoupon(code[i], businessLine[i], isActive[i])) {
                coupons.add(new Coupon(code[i], businessLine[i]));
            }
        }

        Collections.sort(coupons, (a, b) -> {
            int businessLineComparison = Integer.compare(businessLineOrder.get(a.businessLine), businessLineOrder.get(b.businessLine));
            if (businessLineComparison != 0) {
                return businessLineComparison;
            } else {
                return a.code.compareTo(b.code);
            }
        });

        for (Coupon coupon : coupons) {
            validCoupons.add(coupon.code);
        }

        return validCoupons;
    }

    private boolean isValidCoupon(String code, String businessLine, boolean isActive) {
        if (code == null || code.isEmpty()) {
            return false;
        }

        for (char c : code.toCharArray()) {
            if (!Character.isLetterOrDigit(c) && c != '_') {
                return false;
            }
        }

        if (!businessLine.equals("electronics") && !businessLine.equals("grocery") &&
                !businessLine.equals("pharmacy") && !businessLine.equals("restaurant")) {
            return false;
        }

        return isActive;
    }

    private static class Coupon {
        String code;
        String businessLine;

        public Coupon(String code, String businessLine) {
            this.code = code;
            this.businessLine = businessLine;
        }
    }
}

File: problems/3606-coupon-code-validator/solution.py
class Solution:
    def validateCoupons(self, code: List[str], businessLine: List[str], isActive: List[bool]) -> List[str]:
        valid_coupons = []
        n = len(code)
        for i in range(n):
            if isActive[i]:
                if code[i] != "" and code[i].replace("_", "").isalnum():
                    if businessLine[i] in ["electronics", "grocery", "pharmacy", "restaurant"]:
                        valid_coupons.append((businessLine[i], code[i]))

        valid_coupons.sort(key=lambda x: (["electronics", "grocery", "pharmacy", "restaurant"].index(x[0]), x[1]))
        
        return [coupon[1] for coupon in valid_coupons]